modules/k8s/tunnel.go

```diff
@@ -0,0 +1,225 @@
+		return err
+	}
+	kubeConfigPath, err := tunnel.KubectlOptions.GetConfigPath(t)
+	if err != nil {
+		logger.Logf(t, "Error getting kube config path: %s", err)
+		return err
+	}
+	config, err := LoadApiClientConfigE(kubeConfigPath, tunnel.KubectlOptions.ContextName)
+	if err != nil {
+		logger.Logf(t, "Error loading Kubernetes config: %s", err)
+		return err
+	}
+
+	// Find the pod to port forward to
+	podName, err := tunnel.getAttachablePodForResourceE(t)
+	if err != nil {
+		logger.Logf(t, "Error finding available pod: %s", err)
+		return err
+	}
+	logger.Logf(t, "Selected pod %s to open port forward to", podName)
+
+	// Build a url to the portforward endpoint
+	// example: http://localhost:8080/api/v1/namespaces/helm/pods/tiller-deploy-9itlq/portforward
+	postEndpoint := clientset.CoreV1().RESTClient().Post()
+	namespace := tunnel.KubectlOptions.Namespace
+	portForwardCreateURL := postEndpoint.
+		Resource("pods").
+		Namespace(namespace).
+		Name(podName).
+		SubResource("portforward").
+		URL()
+
+	logger.Logf(t, "Using URL %s to create portforward", portForwardCreateURL)
+
+	// Construct the spdy client required by the client-go portforward library
+	transport, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		logger.Logf(t, "Error creating http client: %s", err)
+		return err
+	}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", portForwardCreateURL)
+
+	// Construct a new PortForwarder struct that manages the instructed port forward tunnel
+	ports := []string{fmt.Sprintf("%d:%d", tunnel.LocalPort, tunnel.RemotePort)}
+	portforwarder, err := portforward.New(dialer, ports, tunnel.StopChan, tunnel.ReadyChan, tunnel.Out, tunnel.Out)
+	if err != nil {
+		logger.Logf(t, "Error creating port forwarding tunnel: %s", err)
+		return err
+	}
+
+	// Open the tunnel in a goroutine so that it is available in the background. Report errors to the main goroutine via
+	// a new channel.
+	errChan := make(chan error)
+	go func() {
+		errChan <- portforwarder.ForwardPorts()
+	}()
+
+	// Wait for an error or the tunnel to be ready
+	select {
+	case err = <-errChan:
+		logger.Logf(t, "Error starting port forwarding tunnel: %s", err)
+		return err
+	case <-portforwarder.Ready:
+		logger.Logf(t, "Successfully created port forwarding tunnel")
+		return nil
+	}
+}
+
+// GetAvailablePort retrieves an available port on the host machine. This delegates the port selection to the golang net
+// library by starting a server and then checking the port that the server is using. This will fail the test if it could
+// not find an avilable port.
+func GetAvailablePort(t *testing.T) int {
+	port, err := GetAvailablePortE(t)
+	require.NoError(t, err)
+	return port
+}
+
+// GetAvailablePortE retrieves an available port on the host machine. This delegates the port selection to the golang net
+// library by starting a server and then checking the port that the server is using.
+func GetAvailablePortE(t *testing.T) (int, error) {
```

Comment:
If you run this in a concurrent context, couldn't a race condition lead to the same port being selected? i.e., goroutine one listens on port 12345, it's available, so it returns the port number, and in the `defer` stops listening. Right after the `defer`, but before the port number is returned, another goroutine happens to listen on 12345, see it's available, and returns it.

Reply:
I think you are right. Let me see if there is a way to make this work correctly.

Reply:
Ok resolved this by changing it so that the available port selection happens in `ForwardPortE` when local port is 0. This way, we can enforce a mutex lock between getting the available port and spawning the listener. See 7440db7 .